In [1]:
#Exploratory Data Analysis and Wrangling
import pandas as pd
import numpy as np
import random as rnd

In [2]:
#Data Visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [3]:
#Machine learning
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import precision_score, recall_score, f1_score

In [4]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error 
from math import sqrt
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn import linear_model, datasets

In [5]:
#Read the csv file
df = pd.read_csv("C:/Users/Shreya Sajid/Documents/Git/Alzheimer's Disease Prediction/oasis_longitudinal.csv")

## 1. Data Cleaning

#### Fill the missing values in SES with median:

In [26]:
df["SES"].fillna(df["SES"].median(), inplace=True)

#### Fill the missing values in MMSE with mean:

In [27]:
df["MMSE"].fillna(df["MMSE"].mean(), inplace=True)

## 2. Data Wrangling

##### Retaining only the observations with the first visit of the patients:

In [28]:
df = df.loc[df['Visit']==1] 
df = df.reset_index(drop=True)

##### Dropping the unnecessary columns

In [29]:
df = df.drop(['MRI ID', 'Visit', 'Hand'], axis=1)

##### Convert the Converted group into Demented

In [30]:
df['Group'] = df['Group'].replace(['Converted'], ['Demented'])

In [31]:
df['Group'].value_counts()

Demented       78
Nondemented    72
Name: Group, dtype: int64

##### Label Encoding

In [32]:
le = LabelEncoder()

In [33]:
var_mod = ['Group','M/F']

In [34]:
#Convert to numericals
for i in var_mod:
    df[i] = le.fit_transform(df[i].astype(str))

In [35]:
#CROSS-CHECK
df.dtypes

Subject ID     object
Group           int32
MR Delay        int64
M/F             int32
Age             int64
EDUC            int64
SES           float64
MMSE          float64
CDR           float64
eTIV            int64
nWBV          float64
ASF           float64
dtype: object

In [36]:
df.head()

Subject ID  Group  MR Delay  M/F  Age  EDUC  SES  MMSE  CDR  eTIV   nWBV  \
0  OAS2_0001      1         0    1   87    14  2.0  27.0  0.0  1987  0.696   
1  OAS2_0002      0         0    1   75    12  2.0  23.0  0.5  1678  0.736   
2  OAS2_0004      1         0    0   88    18  3.0  28.0  0.0  1215  0.710   
3  OAS2_0005      1         0    1   80    12  4.0  28.0  0.0  1689  0.712   
4  OAS2_0007      0         0    1   71    16  2.0  28.0  0.5  1357  0.748   

     ASF  
0  0.883  
1  1.046  
2  1.444  
3  1.039  
4  1.293

In [37]:
df = df.drop(['MR Delay','Subject ID','CDR'], axis=1)

## 3. Predicting Using Machine Learning Algorithms

In [38]:
#To split the dataset
from sklearn.model_selection import train_test_split

In [39]:
Y = df.Group
df.drop(['Group'],axis=1,inplace=True)

In [40]:
X_train, X_test, Y_train, Y_test = train_test_split(df, Y, test_size=0.33, random_state=42)

In [41]:
def model_report(y_true,y_pred):
    print("Accuracy = ",accuracy_score(y_true,y_pred))
    print("Recall = ",recall_score(y_true,y_pred))
    print("Precision = ",precision_score(y_true,y_pred))
    print("F1 Score = ",f1_score(y_true,y_pred))
    pass

### 3.1 Logistic Regression

In [42]:
#Fit the model
logreg = LogisticRegression()
logreg.fit(X_train, Y_train)

C:\Users\Shreya Sajid\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression()

In [43]:
#TRAIN SCORE
Y_pred = logreg.predict(X_train)
acc_log = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_log)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.77
RMSE =  0.47958315233127197


In [44]:
#TEST SCORE
Y_pred = logreg.predict(X_test)
acc_log_test = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_log_test)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.6
RMSE =  0.6324555320336759


In [45]:
#Generate the model report
print("MODEL REPORT:")
model_report(Y_test,Y_pred)

MODEL REPORT:
Accuracy =  0.6
Recall =  0.5
Precision =  0.65
F1 Score =  0.5652173913043479


### 3.2 K Nearest Neighbour

In [46]:
#Fit the model
knn = KNeighborsClassifier(n_neighbors = 3) 
knn.fit(X_train, Y_train)

KNeighborsClassifier(n_neighbors=3)

In [47]:
Y_pred = knn.predict(X_train)  
acc_knn = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_knn)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.76
RMSE =  0.4898979485566356


In [48]:
Y_pred = knn.predict(X_test)  
acc_knn_test = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_knn_test)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.56
RMSE =  0.6633249580710799


In [49]:
#Generate the model report
print("MODEL REPORT:")
model_report(Y_test,Y_pred)

MODEL REPORT:
Accuracy =  0.56
Recall =  0.5769230769230769
Precision =  0.5769230769230769
F1 Score =  0.5769230769230769


### 3.3 Support Vector Machines

In [50]:
#Fit the model
svc = SVC()
svc.fit(X_train, Y_train)

SVC()

In [51]:
Y_pred = svc.predict(X_train)

acc_svc = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_svc)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.54
RMSE =  0.6782329983125268


In [52]:
Y_pred = svc.predict(X_test)
acc_svc_test = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_svc_test)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.48
RMSE =  0.7211102550927979


In [53]:
#Generate the model report
print("MODEL REPORT:")
model_report(Y_test,Y_pred)

MODEL REPORT:
Accuracy =  0.48
Recall =  0.0
Precision =  0.0
F1 Score =  0.0


C:\Users\Shreya Sajid\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### 3.4 Decision Tree

In [54]:
#Fit the model
decision_tree = DecisionTreeClassifier() 
decision_tree.fit(X_train, Y_train) 

DecisionTreeClassifier()

In [55]:
Y_pred = decision_tree.predict(X_train)  
acc_decision_tree = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_decision_tree)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  1.0
RMSE =  0.0


In [56]:
Y_pred = decision_tree.predict(X_test)  
acc_decision_tree_test = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_decision_tree_test)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.66
RMSE =  0.5830951894845301


### Hyperparameter Tuning

In [57]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn import decomposition

In [58]:
sc = StandardScaler()
pca = decomposition.PCA()

pipe = Pipeline(steps=[('sc', sc),
                           ('pca', pca),
                           ('decision_tree', decision_tree)])

n_components = list(range(1,X_train.shape[1]+1,1))


criterion = ['gini', 'entropy']
max_depth = [4,6,8,12]

parameters = dict(pca__n_components=n_components,
                      decision_tree__criterion=criterion,
                      decision_tree__max_depth=max_depth)

decision_tree = GridSearchCV(pipe, parameters)

In [59]:
decision_tree.fit(X_train,Y_train)

GridSearchCV(estimator=Pipeline(steps=[('sc', StandardScaler()), ('pca', PCA()),
                                       ('decision_tree',
                                        DecisionTreeClassifier())]),
             param_grid={'decision_tree__criterion': ['gini', 'entropy'],
                         'decision_tree__max_depth': [4, 6, 8, 12],
                         'pca__n_components': [1, 2, 3, 4, 5, 6, 7, 8]})

In [60]:
#View The Best Parameters
print('Best Criterion:', decision_tree.best_estimator_.get_params()['decision_tree__criterion'])
print('Best max_depth:', decision_tree.best_estimator_.get_params()['decision_tree__max_depth'])
print('Best Number Of Components:', decision_tree.best_estimator_.get_params()['pca__n_components'])
print(); print(decision_tree.best_estimator_.get_params()['decision_tree'])

Best Criterion: gini
Best max_depth: 6
Best Number Of Components: 6

DecisionTreeClassifier(max_depth=6)


In [61]:
#Fit the model
decision_tree = DecisionTreeClassifier(max_depth = 6) 
decision_tree.fit(X_train, Y_train) 

DecisionTreeClassifier(max_depth=6)

In [62]:
Y_pred = decision_tree.predict(X_train)  
acc_decision_tree = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_decision_tree)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.98
RMSE =  0.1414213562373095


In [63]:
Y_pred = decision_tree.predict(X_test)  
acc_decision_tree_test = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_decision_tree_test)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.64
RMSE =  0.6


In [64]:
#Generate the model report
print("MODEL REPORT:")
model_report(Y_test,Y_pred)

MODEL REPORT:
Accuracy =  0.64
Recall =  0.5769230769230769
Precision =  0.6818181818181818
F1 Score =  0.6249999999999999


### 3.5 XGBoost Classifier

In [65]:
import xgboost as xgb
from xgboost import XGBClassifier

In [66]:
#Fit the model
xgb = XGBClassifier(learning_rate = 0.05, n_estimators=100, max_depth=5)
xgb.fit(X_train,Y_train)

XGBClassifier(learning_rate=0.05, max_depth=5)

In [67]:
Y_pred = xgb.predict(X_train)
acc_xgb = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_xgb)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.98
RMSE =  0.1414213562373095


In [68]:
Y_pred = xgb.predict(X_test)
acc_xgb = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_xgb)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.68
RMSE =  0.565685424949238


### Hyperparameter Tuning

In [69]:
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import StratifiedKFold

In [70]:
params = {
    'min_child_weight': [1, 5, 10],
        'gamma': [0.5, 1, 1.5, 2, 5],
        'subsample': [0.6, 0.8, 1.0],
        'colsample_bytree': [0.6, 0.8, 1.0],
        'max_depth': [3, 4, 5]
        }

In [71]:
xgb = XGBClassifier(learning_rate=0.02, n_estimators=600, objective='binary:logistic',
                    silent=True, nthread=1)

In [72]:
folds = 3
param_comb = 5

skf = StratifiedKFold(n_splits=folds, shuffle = True, random_state = 1001)

random_search = RandomizedSearchCV(xgb, param_distributions=params, n_iter=param_comb, scoring='roc_auc', n_jobs=4, cv=skf.split(X_train,Y_train), verbose=3, random_state=1001 )


In [73]:
random_search.fit(X_train, Y_train)

Fitting 3 folds for each of 5 candidates, totalling 15 fits


[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  15 out of  15 | elapsed:   12.9s finished


RandomizedSearchCV(cv=<generator object _BaseKFold.split at 0x00000278B97C9548>,
                   estimator=XGBClassifier(learning_rate=0.02, n_estimators=600,
                                           nthread=1, silent=True),
                   n_iter=5, n_jobs=4,
                   param_distributions={'colsample_bytree': [0.6, 0.8, 1.0],
                                        'gamma': [0.5, 1, 1.5, 2, 5],
                                        'max_depth': [3, 4, 5],
                                        'min_child_weight': [1, 5, 10],
                                        'subsample': [0.6, 0.8, 1.0]},
                   random_state=1001, scoring='roc_auc', verbose=3)

In [74]:
print('\n All results:')
print(random_search.cv_results_)
print('\n Best estimator:')
print(random_search.best_estimator_)
print('\n Best normalized gini score for %d-fold search with %d parameter combinations:' % (folds, param_comb))
print(random_search.best_score_ * 2 - 1)
print('\n Best hyperparameters:')
print(random_search.best_params_)
results = pd.DataFrame(random_search.cv_results_)


 All results:
{'mean_fit_time': array([1.46467392, 1.78073279, 0.92429701, 1.06613413, 1.45505182]), 'std_fit_time': array([0.08250866, 0.25669982, 0.07211003, 0.00729139, 0.16920863]), 'mean_score_time': array([0.01326744, 0.00786471, 0.01075737, 0.00563971, 0.00735418]), 'std_score_time': array([0.00810563, 0.00140226, 0.00362324, 0.00111563, 0.00073281]), 'param_subsample': masked_array(data=[1.0, 0.6, 0.8, 1.0, 0.8],
             mask=[False, False, False, False, False],
       fill_value='?',
            dtype=object), 'param_min_child_weight': masked_array(data=[5, 1, 5, 5, 1],
             mask=[False, False, False, False, False],
       fill_value='?',
            dtype=object), 'param_max_depth': masked_array(data=[3, 5, 5, 5, 4],
             mask=[False, False, False, False, False],
       fill_value='?',
            dtype=object), 'param_gamma': masked_array(data=[5, 1.5, 1, 5, 1],
             mask=[False, False, False, False, False],
       fill_value='?',
            dt

In [75]:
#Fit the model
xgb = XGBClassifier(subsample = 0.8, min_child_weight = 1, max_depth = 4, gamma = 1, colsample_bytree = 1.0)
xgb.fit(X_train,Y_train)

XGBClassifier(colsample_bytree=1.0, gamma=1, max_depth=4, subsample=0.8)

In [76]:
Y_pred = xgb.predict(X_train)
acc_xgb = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_xgb)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.94
RMSE =  0.2449489742783178


In [77]:
Y_pred = xgb.predict(X_test)
acc_xgb = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_xgb)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.74
RMSE =  0.5099019513592785


In [78]:
#Generate the model report
print("MODEL REPORT:")
model_report(Y_test,Y_pred)

MODEL REPORT:
Accuracy =  0.74
Recall =  0.7692307692307693
Precision =  0.7407407407407407
F1 Score =  0.7547169811320754


### 3.6 Random Forest

In [79]:
#Fit the model
random_forest = RandomForestClassifier(n_estimators=100,min_samples_split=25,max_depth=7,max_features=1)
random_forest.fit(X_train, Y_train)

RandomForestClassifier(max_depth=7, max_features=1, min_samples_split=25)

In [80]:
Y_pred = random_forest.predict(X_train)

acc_random_forest = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_random_forest)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.84
RMSE =  0.4


In [81]:
Y_pred = random_forest.predict(X_test)
acc_random_forest_test = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_random_forest_test)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.68
RMSE =  0.565685424949238


### Cross Validation

##### We will perform K-fold cross validation with 10 folds to output an array of 10 results.

In [82]:
scores = cross_val_score(random_forest, X_train, Y_train, cv=10, scoring = "accuracy")
print("Train Scores:", scores)
print("Mean:", scores.mean())
print("Standard Deviation:", scores.std())

Train Scores: [0.7 0.9 0.5 0.9 0.7 0.6 0.6 0.7 0.6 0.7]
Mean: 0.69
Standard Deviation: 0.12206555615733705


In [83]:
scores = cross_val_score(random_forest, X_test, Y_test, cv=10, scoring = "accuracy")
print("Test Scores:", scores)
print("Mean:", scores.mean())
print("Standard Deviation:", scores.std())

Test Scores: [1.  0.8 0.8 1.  0.8 0.8 0.6 0.6 0.8 0.6]
Mean: 0.78
Standard Deviation: 0.14


### Hyperparameter Tuning

In [84]:
#Create param grid object 
forest_params = dict(     
    max_depth = [n for n in range(9, 14)],     
    min_samples_split = [n for n in range(4, 11)], 
    min_samples_leaf = [n for n in range(2, 5)],     
    n_estimators = [n for n in range(10, 60, 10)],
)

In [85]:
#Instantiate Random Forest model
forest = RandomForestClassifier()

In [86]:
#Build and fit model 
forest_cv = GridSearchCV(estimator=forest, param_grid=forest_params, cv=5) 
forest_cv.fit(X_train, Y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'max_depth': [9, 10, 11, 12, 13],
                         'min_samples_leaf': [2, 3, 4],
                         'min_samples_split': [4, 5, 6, 7, 8, 9, 10],
                         'n_estimators': [10, 20, 30, 40, 50]})

In [87]:
print("Best score: {}".format(forest_cv.best_score_))
print("Optimal params: {}".format(forest_cv.best_estimator_))

Best score: 0.79
Optimal params: RandomForestClassifier(max_depth=13, min_samples_leaf=2, min_samples_split=10,
                       n_estimators=20)


In [88]:
Y_pred = forest_cv.predict(X_train)
acc_random_forest = accuracy_score(Y_train,Y_pred)
print("Train Score = ",acc_random_forest)

#RMSE
rmse = sqrt(mean_squared_error(Y_train, Y_pred))
print("RMSE = ",rmse)

Train Score =  0.93
RMSE =  0.2645751311064591


In [89]:
Y_pred = forest_cv.predict(X_test)
acc_random_forest_test = accuracy_score(Y_test,Y_pred)
print("Test Score = ",acc_random_forest_test)

#RMSE
rmse = sqrt(mean_squared_error(Y_test, Y_pred))
print("RMSE = ",rmse)

Test Score =  0.6
RMSE =  0.6324555320336759


In [90]:
#Generate the model report
print("MODEL REPORT:")
model_report(Y_test,Y_pred)

MODEL REPORT:
Accuracy =  0.6
Recall =  0.6153846153846154
Precision =  0.6153846153846154
F1 Score =  0.6153846153846154


### 4. Best Model

### 4.1 Train Scores

In [93]:
results = pd.DataFrame({
    'Model': ['Logistic Regression', 'KNN','Support Vector Machines','Decision Tree','Random Forest','XGBoost'],
    'Score': [acc_log, acc_knn, acc_svc, acc_decision_tree, acc_random_forest,acc_xgb]})
df_result = results.sort_values(by='Score', ascending=False)
#Display
df_result

Model  Score
3            Decision Tree   0.98
4            Random Forest   0.93
0      Logistic Regression   0.77
1                      KNN   0.76
5                  XGBoost   0.74
2  Support Vector Machines   0.54

### 4.2 Test Scores

In [94]:
results = pd.DataFrame({
    'Model': ['Logistic Regression', 'KNN','Support Vector Machines','Decision Tree','Random Forest','XGBoost'],
    'Score': [acc_log_test, acc_knn_test, acc_svc_test, acc_decision_tree_test, acc_random_forest_test,acc_xgb]})
df_result = results.sort_values(by='Score', ascending=False)
#Display
df_result


Model  Score
5                  XGBoost   0.74
3            Decision Tree   0.64
0      Logistic Regression   0.60
4            Random Forest   0.60
1                      KNN   0.56
2  Support Vector Machines   0.48

### Inference:
From what has been observed, XGBoost is the best model followed by Random Forest along with the appropriate hyperparameter tuning. XGBoost is the best model as it is consistent in both train and test sets compared to other models.

XGBoost with optimal hyperparameter tuning happens to be the best model with a train score of 0.74, a test score also of 0.74, a recall of 0.7504 and an F1 score of 0.76.